core/src/test/java/io/grpc/ContextTest.java

```diff
@@ -0,0 +1,239 @@
+    public void run() {
+      observed = OBSERVED.get();
+      observableLatch.countDown();
+    }
+  };
+
+  private ExecutorService executorService = Executors.newCachedThreadPool();
+
+  @Before
+  public void setUp() throws Exception {
+    // Detach all contexts back to the root.
+    while (Context.current() != Context.ROOT) {
+      Context.current().detach();
+    }
+  }
+
+  @After
+  public void tearDown() throws Exception {
+  }
+
+  public void rootIsInitialContext() {
+    assertNotNull(Context.ROOT);
+    assertSame(Context.ROOT, Context.current());
+  }
+
+  public void rootIsAlwaysBound() {
+    Context root = Context.current();
+    root.detach();
+    assertSame(root, Context.current());
+    root.detach();
+    assertSame(root, Context.current());
+  }
+
+  public void rootCanBeAttached() {
+    Context root = Context.current();
+    Context next = root.withNoCascade();
+    next.attach();
+    root.attach();
+    assertSame(root, Context.current());
+    root.detach();
+    assertSame(next, Context.current());
+    next.detach();
+    assertSame(root, Context.current());
+  }
+
+  // Discussion point. Do we care about strictly throwing or can we be lenient
+  public void rootCannotBeClosed() {
+    assertFalse(Context.ROOT.close(null));
+  }
+
+  public void detachingNonCurrentThrowsIllegalStateException() {
+    Context base = Context.current().withNoCascade();
+    try {
+      base.detach();
+    } catch (IllegalStateException ise) {
+      return;
+    }
+    fail("Expected exception");
+  }
+
+  public void detachUnwindsAttach() {
+    Context base = Context.current().withNoCascade();
+    Context child = base.withNoCascade();
+    Context grandchild = child.withNoCascade();
+    base.attach();
+    base.attach();
+    child.attach();
+    base.attach();
+    grandchild.attach();
+    assertSame(grandchild, Context.current());
+    grandchild.detach();
+    assertSame(base, Context.current());
+    base.detach();
+    assertSame(child, Context.current());
+    child.detach();
+    assertSame(base, Context.current());
+    base.detach();
+    assertSame(base, Context.current());
+    base.detach();
+    assertSame(Context.ROOT, Context.current());
+  }
+
+  public void valuesAndOverrides() {
+    Context base = Context.current().withValue(PET, "dog");
+    Context child = base.withValues(PET, null, FOOD, "cheese");
+
+    base.attach();
+
+    assertEquals("dog", PET.get());
+    assertEquals("lasagna", FOOD.get());
+    assertNull(COLOR.get());
+
+    child.attach();
+
+    assertNull(PET.get());
+    assertEquals("cheese", FOOD.get());
+    assertNull(COLOR.get());
+
+    child.detach();
+
+    // Should have values from base
+    assertEquals("dog", PET.get());
+    assertEquals("lasagna", FOOD.get());
+    assertNull(COLOR.get());
+
+    base.detach();
+
+    assertNull(PET.get());
+    assertEquals("lasagna", FOOD.get());
+    assertNull(COLOR.get());
+  }
+
+  public void doubleCloseIsSilentlyIgnored() {
```

Comment:
I think the test name is misleading since it isn't ignored but communicated via return value. Maybe `closeReturnsFalseIfAlreadyClosed`.